wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,259 @@
+use crate::TagSet;
+
+/// The identifier for the disallowed declaration name rule.
+const ID: &str = "DisallowedDeclarationName";
+
+/// Diagnostic for declaration identifiers with type prefixes or suffixes
+fn decl_identifier_with_type(span: Span, decl_name: &str, type_name: &str) -> Diagnostic {
+    Diagnostic::note(format!("declaration identifier '{}' contains type name '{}'", decl_name, type_name))
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_fix("rename the identifier to not include the type name")
+}
+
+/// A rule that checks for declaration names that contain their type.
+#[derive(Debug, Default)]
+pub struct DisallowedDeclarationNameRule;
+
+impl Visitor for DisallowedDeclarationNameRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn bound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &BoundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(state, &Decl::Bound(decl.clone()), &self.exceptable_nodes());
+        }
+    }
+
+    fn unbound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &UnboundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(
+                state,
+                &Decl::Unbound(decl.clone()),
+                &self.exceptable_nodes(),
+            );
+        }
+    }
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Declaration names should not include their type"
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Declaration names should not include their type as a prefix or suffix. \
+        This makes the code more verbose and often redundant. For example, use \
+        'input_file' instead of 'file_input' or 'myFile'."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::OutputSectionNode,
+        ])
+    }
+}
+
+/// Check declaration name for type prefixes or suffixes
+fn check_decl_name(
+    state: &mut Diagnostics,
+    decl: &Decl,
+    exceptable_nodes: &Option<&'static [SyntaxKind]>,
+) {
+    let name = decl.name();
+    let name_str = name.as_str();
+    let name_lower = name_str.to_lowercase();
+    
+    // Get the declaration type
+    let ty = decl.ty();
+    
+    // Extract base type name(s) using AST methods
+    let mut type_names = Vec::new();
+    
+    // Add the full type name
+    type_names.push(ty.to_string());
```

Comment:
This won't be useful for any compound types, as it will add a string with `[]` brackets in it which aren't valid in an `Identifier`